### ISSUE TITLE
[query/suggestion] Choosing sampled data points in Adaboost

### ISSUE THREAD

**Issue by SiluPanda** [aimacode/aima-python#1047]
Are we using sampled weights for hypothesis training in [AdaBoost](https://github.com/aimacode/aima-python/blob/5e5b51c47a6573c636c424e67bcae8dbbbebc87e/learning.py#L935-L958)? If not should not we be doing that? 
Sampled weight means, choosing a set of data points from the training set according to their weights. It increases the chance of a misclassified point to get chosen with a greater probability by the next hypothesis.
Here is the implementation of [AdaBoost](https://github.com/SiluPanda/ensemble-learning-bagging-and-boosting/blob/master/ensembling/boosting.py) done by me. 
I have implemented sampling of data [here](https://github.com/SiluPanda/ensemble-learning-bagging-and-boosting/blob/a211a7467f619d2b565e6b66a2f565dc22ef11d3/ensembling/perceptron.py#L78-L91).

Let me know if needs to be done.

**Comment by MrDupin** [comment]
Hmm, this looks very interesting, but I am not sure if the AdaBoost implementation is needed at the moment. For that, only @norvig can respond, but he is very busy at the moment. If you are doing this for GSoC, you can add in your proposal that you will do some work on neural networks. I think that would be a really interesting idea, but as a PR, it would be too large for me to merge.

**Comment by SiluPanda** [comment]
Yes, Thank you, I'll add that on my proposal. 
Another small query, what is exaclty `size` [here](https://github.com/aimacode/aima-python/blob/5e5b51c47a6573c636c424e67bcae8dbbbebc87e/learning.py#L1089-L1117)? I am looking to patch the infinite loop.
Thanks a lot for the response!

**Comment by MrDupin** [comment]
To be honest, I don't know. The cross validation pseudocode is not up to data and we don't know what to do.

**Comment by SiluPanda** [comment]
This what `size` means from the book:

> In this section we explain how to select among models that are parameterized by size.
For example, with polynomials we have size = 1 for linear functions, size = 2 for quadratics,
and so on. For decision trees, the size could be the number of nodes in the tree. In all cases
we want to find the value of the size parameter that best balances underfitting and overfitting
to give the best test set accuracy.

Clearly, `size` should not go till infinity(as it is in the pseudocode) and the upper limit to it is model specific. I guess the best option is to wait for a update on the source code from @norvig sir.

**Comment by dmeoli** [comment]
Weighted sampling is already supported: `ada_boost(dataset, L, K)` is driven by a weighted learner, and the library provides `WeightedLearner` → `replicated_dataset` → `weighted_replicate` (in `learning.py`), which resamples the examples in proportion to their weights — exactly the 'choosing data points according to their weights' you describe. So no change is needed to train on weight-sampled data. Closing as resolved.
